lib/icingadb/redisconnection.hpp

```diff
@@ -387,6 +394,67 @@ void RedisConnection::WriteOne(StreamPtr& stream, RedisConnection::Query& query,
 	}
 }
 
+/**
+ * Initialize a Redis stream
+ *
+ * @param stream Redis server connection
+ * @param query Redis query
+ */
+template<class StreamPtr>
+void RedisConnection::Handshake(StreamPtr& stream, boost::asio::yield_context& yc)
+{
+	auto strm (stream);
+
+	if (m_Password.IsEmpty() && !m_DbIndex) {
+		// Trigger NOAUTH
+		WriteRESP(*strm, {"PING"}, yc);
+	} else {
+		if (!m_Password.IsEmpty()) {
+			WriteRESP(*strm, {"AUTH", m_Password}, yc);
+		}
+
+		if (m_DbIndex) {
+			WriteRESP(*strm, {"SELECT", Convert::ToString(m_DbIndex)}, yc);
+		}
+	}
+
+	strm->async_flush(yc);
+
+	if (m_Password.IsEmpty() && !m_DbIndex) {
+		Reply pong (ReadRESP(*strm, yc));
+
+		if (pong.IsObjectType<RedisError>()) {
+			// Likely NOAUTH
+			BOOST_THROW_EXCEPTION(std::runtime_error(RedisError::Ptr(pong)->GetMessage()));
+		}
+	} else {
+		if (!m_Password.IsEmpty()) {
+			Reply auth (ReadRESP(*strm, yc));
+
+			if (auth.IsObjectType<RedisError>()) {
+				auto& authErr (RedisError::Ptr(auth)->GetMessage().GetData());
+				boost::smatch what;
+
+				if (boost::regex_search(authErr, what, m_ErrAuth)) {
```

Comment:
Is `ERR AUTH <password> called without any password configured for the default user. Are you sure your configuration is correct?`` really the only error starting with `ERR AUTH `? So is it safe to not treat this as an error in general?

IMHO it would also be fine to treat this as a misconfiguration and bail out, eliminating the need for this check altogether.

Should we keep this, do we really have to use regex here? Shouldn't this do the very same as `if (authErr.Find("ERR AUTH ") == 0)`?

Reply:
1. Yes.
2. This misconfiguration doesn’t hurt anything.
3. Regex should give up the search earlier if the subject is missing.

Reply:
1. Do you have a source for this?
2. Yes, but only if you can reliably detect that it's this very configuration.
3. Must be a really long error message that invoking a regex engine is cheaper that iterating the string once, but fine if you insist on it.

Reply:
1. Well, the only other AUTH error is a distinct one, WRONGPASS.
2. If you can... what?
3. A good regex engine jumps to `\A` (O(1)) and check character-by-character. Once one unexpected character appears, it gives up. (O(common_prefix(N, M)))

Reply:
1. Is there any documentation on which errors may be returned? If there was another one starting with `"ERR AUTH "` (or was added in the future) the code would incorrectly assume that this was a superfluous password.
2. The misconfiguration itself wouldn't hurt, but it would hurt if what I described in (1) could happen.
3. The initial reason for my comment wasn't even about performance. What I had in mind was to improve the readability of the code by avoiding the indirection (`"ERR AUTH "` appears right where it's used and you don't have to follow the indirection of the `m_ErrAuth` variable).

Side note: your comment shows a common misconception about asymptotic runtime: it does not say that an O(1) implementation is always faster than O(N) but only that you can find a sufficiently large N so that the O(1) implementation becomes faster than O(N).